File: scala-playground/src/main/scala/functional/Either.scala
package functional

// Define the usual Either API.
//
// * Constructors (on the object)
// left
// right
// * methods
// map
// flatMap
// orElse
// getOrElse
// any other useful functions

sealed trait Either[+A, +B] {

  def fold[X](left: A => X, right: B => X): X = this match {
    case Left(v) => left(v)
    case Right(v) => right(v)
  }

  def map[Y](f: B => Y): Either[A, Y] = this match {
    case l@Left(_) => l
    case Right(v) => Right(f(v))
  }

  def flatMap[AA >: A, Y](f: B => Either[AA, Y]): Either[AA, Y] = this match {
    case l@Left(_) => l
    case Right(v) => f(v)
  }

  def orElse[AA >: A, Y](f: => Either[AA, Y]): Either[AA, Y] = flatMap(_ => f)

  def getOrElse[C >: B](f: => C): C = fold(_ => f, a => a)

  def swap: Either[B, A] = this match {
    case Left(v) => Right(v)
    case Right(v) => Left(v)
  }

}

final case class Left[A](value: A) extends Either[A, Nothing]
final case class Right[B](value: B) extends Either[Nothing, B]

File: funkotlin/src/main/kotlin/funkotlin/Either.kt
package funkotlin

sealed class Either<out L, out R> {

    fun <X> fold(f: (L) -> X, g: (R) -> X) = when (this) {
        is Left -> f(l)
        is Right -> g(r)
    }

    val isLeft: Boolean = fold({ true }, { false })
    val isRight: Boolean = !isLeft

    fun left() = LeftProjection(this)

}

data class Left<out L, out R>(val l: L) : Either<L, R>()
data class Right<out L, out R>(val r: R) : Either<L, R>()

data class LeftProjection<out L, out R>(val either: Either<L, R>) {

    fun <X> map(f: (L) -> X): LeftProjection<X, R> = flatMap { Left<X, R>(f(it)) }

}

fun <L, R, X> LeftProjection<L, R>.flatMap(f: (L) -> Either<X, R>): LeftProjection<X, R> = when (either) {
    is Left -> LeftProjection(f(either.l))
    is Right -> LeftProjection(Right(either.r))
}

data class RightProjection<out L, out R>(val either: Either<L, R>) {

    fun <X> map(f: (R) -> X): RightProjection<L, X> = flatMap { Right<L, X>(f(it)) }

}

fun <L, R, X> RightProjection<L, R>.flatMap(f: (R) -> Either<L, X>): RightProjection<L, X> = when (either) {
    is Left -> RightProjection(Left(either.l))
    is Right -> RightProjection(f(either.r))
}